### ISSUE TITLE
Combine and zip doesn't work if one of the flows doesn't emmit value

### ISSUE THREAD

**Issue by Y2Kot** [Kotlin/kotlinx.coroutines#2667]
Here is description in code:

This code works good and return correct lists for both combine and zip methods
```
fun main() = runBlocking {
    val flowOfInt: Flow<Int> = flow {
        emit(1)
        delay(100)
        emit(2)
        delay(100)
        emit(3)
        delay(100)
    }
    val flowOfNull = flow<Int> {
        emit(4)
        delay(50)
        emit(5)
        delay(50)
        emit(6)
        delay(50)
    }

    val combine = flowOfInt.zip(flowOfNull) { a, b ->
        listOf(a, b)
    }
    combine.collect { println(it) }
}
```
result:
```
[1, 4]
[2, 5]
[3, 6]
```

However if one of the flows doesn't emmit, both combines and zip stop working:
```
fun main() = runBlocking {
    val flowOfInt: Flow<Int> = flow {
        emit(1)
        delay(100)
        emit(2)
        delay(100)
        emit(3)
        delay(100)
    }
    val flowOfNull = flow<Int> {}

    val combine = flowOfInt.zip(flowOfNull) { a, b ->
        listOf(a, b)
    }
    combine.collect { println(it) }
}
```

result:
` `

Sometimes it's really needed to get at list one value.
Possibly it can be a good variant to add methods like:
`combineAny{ a?, b? -> }`
and
`zipAny{ a?, b? -> }`

**Comment by dkhalanskyjb** [comment]
Yeah, semantically, `zip` is a method that attempts to build as many pairs as possible of corresponding elements of two streams, so if some stream finishes early and pairs can no longer be matched, `zip` just stops. This is a convention that goes beyond Kotlin and is employed in many other languages, so this behavior is expected.

It's not clear when one would need to build pairs and then also access the elements that couldn't be matched. Typically, if the task is just acquiring elements from two streams in parallel, without the need to combine them into pairs, `merge` should suffice. Could you please share your specific use cases for `zipAny`?

**Comment by Y2Kot** [comment]
I'm collecting system sensors data to determine the user's state (falling, fatigue etc) and showing sensors state to the screen. I'm using callbackflow to convert liseners to data flow and later in repository combine values into list of sensors values. When one of sensor (for example pressure) is missing, I don't getting any data at all.

**Comment by dkhalanskyjb** [comment]
Interesting, thanks! Is it true that all the streams you're dealing with are either infinite or empty? Or can a sensor become missing during the execution of the code?

**Comment by Y2Kot** [comment]
Yes, example:
I have 3 sensors for the whole app lifetime:
Magnetic(all devices have it) send data very often - for example each 200ms 
Pressure (mostly flagship devices have it) send data very often - for example each 200ms if sensor exists
Battery (all devices have it) frequency mostly depends on vendor (some devices send data as you setup it and some send data on change).

**Comment by dkhalanskyjb** [comment]
I see, this makes sense. It does look like `combine` is a suitable operator in this scenario.

Would your use case be covered by using `flowOf(null)` instead of `flow { }` to represent the lack of readings? This way, `combine` will always be returning something, with the latest reading from the missing sensors always being `null`.

Something like this:
```kotlin
fun <T> Flow<T>.prependNull(): Flow<T?> = flow {
    emit(null)
    emitAll(this@prependNull)
}

val flow1 = flow {
    repeat(3) {
        delay(200)
        emit(it)
    }
}
val flow2 = flow<Int> { }.prependNull()
assertEquals(listOf(Pair(0, null), Pair(1, null), Pair(2, null)),
    flow1.combine(flow2) { a, b -> Pair(a, b) }.toList())
}
```

**Comment by Y2Kot** [comment]
yes my first solution was to make 
```
callbackflow<SensorData?> {
    offer(null)
    listenerHere() {}
}
```

However its not looking nice, so I found another solution:
```
@OptIn(ExperimentalCoroutinesApi::class)
inline fun <reified T> instantCombine(vararg flows: Flow<T>) = channelFlow {
    val array= Array(flows.size) {
        false to (null as T?) // first element stands for "present"
    }

    flows.forEachIndexed { index, flow ->
        launch {
            flow.collect { emittedElement ->
                array[index] = true to emittedElement
                send(array.filter { it.first }.mapNotNull { it.second })
            }
        }
    }
}
```
that was got here - https://stackoverflow.com/questions/61185082/combine-multiple-kotlin-flows-in-a-list-without-waiting-for-a-first-value

However still there are some problems with nullability - send(array.filter { it.first }.mapNotNull { it.second }) (difference with original answer)
System sensors cant produce null values (the can not exists but its another story)

**Comment by dkhalanskyjb** [comment]
My main point is that it looks like `combineAny` as defined above could be implemented ad-hoc as follows:
```kotlin
fun <T> Flow<T>.prependNull(): Flow<T?> = flow {
    emit(null)
    emitAll(this@prependNull)
}

fun <T1, T2, R> Flow<T1>.combineAny(flow: Flow<T2>, transform: suspend (a: T1?, b: T2?) -> R): Flow<R> =
    prependNull().combine(flow.prependNull(), transform)
```
This way, you don't even have to change what you store in the repository, so the property of not looking nice is limited to the implementation of `combineAny`. An additional bonus of such ad-hoc implementation is that if, for example, one knows that `Flow<T1>` always outputs some values, then the first `prependNull` is not needed, and `transform` can accept `T1` and not `T1?`.

Regarding the code from StackOverflow: it does deal with the problem, but it is limited to the cases where all the streams produce the values of the same type and it doesn't matter which values came from which stream.

**Comment by Y2Kot** [comment]
I've tested your variant for some time and it's really more flexible and good looking

**Comment by qwwdfsad** [comment]
In the very same way, you have to prepend `null` to each flow
```
inline fun <reified T, R> combineAny(vararg flows: Flow<T>, crossinline transform: suspend (Array<T?>) -> R): Flow<R> =
            combine(flows.map { it.prependNull() }, transform)
```
